pkg/operator/migration_platform_status/controller.go

```diff
@@ -75,41 +72,72 @@ func (c MigrationAWSStatusController) sync(ctx context.Context, syncCtx factory.
 
 	currentInfra := obji.DeepCopy()
 
-	if currentInfra.Status.Platform != configv1.AWSPlatformType || // not aws
-		(currentInfra.Status.PlatformStatus != nil && currentInfra.Status.PlatformStatus.Type != configv1.AWSPlatformType) || // not aws
-		(currentInfra.Status.PlatformStatus != nil && currentInfra.Status.PlatformStatus.Type == configv1.AWSPlatformType &&
-			currentInfra.Status.PlatformStatus.AWS != nil && len(currentInfra.Status.PlatformStatus.AWS.Region) > 0) { // aws, but the region is already set. so no need to migrate
-		return nil // no action
+	if currentInfra.Status.PlatformStatus == nil {
+		currentInfra.Status.PlatformStatus = &configv1.PlatformStatus{}
+	}
+	if currentInfra.Status.PlatformStatus.Type == "" {
+		currentInfra.Status.PlatformStatus.Type = currentInfra.Status.Platform
+	}
+	if old, new := currentInfra.Status.Platform, currentInfra.Status.PlatformStatus.Type; old != "" && new != "" && old != new {
+		message := fmt.Sprintf("Mis-match between status.platform (%s) and status.platformStatus.type (%s) in infrastructures.%s/cluster", old, new, configv1.GroupName)
+		syncCtx.Recorder().Warningf("MigrationPlatformStatusController", message)
+		return fmt.Errorf(message)
+	}
+
+	if err := c.migratePlatformSpecificFields(ctx, currentInfra); err != nil {
+		syncCtx.Recorder().Warningf("MigrationPlatformStatusController", err.Error())
+		return err
+	}
+
+	if reflect.DeepEqual(obji.Status.PlatformStatus, currentInfra.Status.PlatformStatus) {
+		// no changes made to platform status
+		return nil
+	}
+
+	_, err = c.infraClient.UpdateStatus(ctx, currentInfra, metav1.UpdateOptions{})
+	if err != nil {
+		syncCtx.Recorder().Warningf("MigrationPlatformStatusController", "Unable to update the infrastructure status")
+		return err
+	}
+	return nil
+}
+
+func (c MigrationPlatformStatusController) migratePlatformSpecificFields(ctx context.Context, currentInfra *configv1.Infrastructure) error {
+	if currentInfra.Status.PlatformStatus.Type != configv1.AWSPlatformType {
+		// only AWS has platform-specific fields to migrate
+		return nil
+	}
+
+	if currentInfra.Status.PlatformStatus.AWS == nil {
+		currentInfra.Status.PlatformStatus.AWS = &configv1.AWSPlatformStatus{}
+	}
+
+	if currentInfra.Status.PlatformStatus.AWS.Region != "" {
+		// region is already set, so no need to migrate
+		return nil
 	}
 
 	cc, err := loadClusterConfig(ctx, c.configMapClient)
 	if err != nil {
-		syncCtx.Recorder().Warningf("MigrationAWSStatusController", "Unable to load the cluster-config-v1")
```

Comment:
this still seems event worthy

Reply:
This error will be recorded as an event by the caller of `migratePlatformSpecificFields`.